File: app/schemas/eligibility.py
from datetime import date
from typing import Any
from pydantic import BaseModel, Field


class EligibilityCheckRequest(BaseModel):
    date_of_birth: date | None = Field(
        None,
        examples=["1985-06-20"],
        description="Date of birth in YYYY-MM-DD format (used to compute age)",
    )
    age: int | None = Field(
        None,
        ge=0,
        le=125,
        examples=[40],
        description="Citizen age in years (alternative to date_of_birth)",
    )
    gender: str | None = Field(
        None,
        examples=["female"],
        description="Gender of applicant (female, male, other)",
    )
    state: str | None = Field(
        None,
        examples=["Maharashtra"],
        description="State of residency",
    )
    district: str | None = Field(
        None,
        examples=["Pune"],
        description="District of residency",
    )
    annual_income: int | None = Field(
        None,
        ge=0,
        examples=[120000],
        description="Annual family income in INR",
    )
    occupation: str | None = Field(
        None,
        examples=["farmer"],
        description="Primary occupation or trade",
    )
    caste_category: str | None = Field(
        None,
        examples=["General", "OBC", "SC", "ST", "EWS"],
        description="Social / caste reservation category",
    )
    is_differently_abled: bool | None = Field(
        None,
        examples=[False],
        description="Whether applicant is a person with disability (Divyangjan)",
    )
    marital_status: str | None = Field(
        None,
        examples=["Single", "Married", "Widowed / Single Mother"],
        description="Marital status of applicant",
    )
    residence_area: str | None = Field(
        None,
        examples=["Rural", "Urban"],
        description="Area of residence (Rural / Gramin vs Urban / Nagar)",
    )
    has_land: bool | None = Field(
        None,
        examples=[True],
        description="Whether applicant or household owns agricultural land",
    )


class CriterionVerdict(BaseModel):
    field: str = Field(..., examples=["annual_income"])
    criterion_title: str = Field(..., examples=["Annual Family Income"])
    status: str = Field(..., examples=["passed"], description="'passed' | 'failed' | 'missing_info'")
    your_value: Any = Field(..., examples=["₹120,000"])
    required_condition: str = Field(..., examples=["Maximum ₹200,000 per year"])
    reason: str = Field(
        ...,
        examples=["Your annual income (₹120,000) is within the allowable limit (Maximum ₹200,000 per year)."],
    )


class SchemeExplanation(BaseModel):
    scheme_id: int = Field(..., examples=[1])
    scheme_name: str = Field(..., examples=["Pradhan Mantri Kisan Samman Nidhi"])
    scheme_slug: str = Field(..., examples=["pm-kisan"])
    ministry: str = Field(..., examples=["Ministry of Agriculture and Farmers Welfare"])
    description: str = Field(..., examples=["Direct income support of ₹6,000 per year..."])
    status: str = Field(..., examples=["eligible"], description="'eligible' | 'nearly_eligible' | 'ineligible'")
    is_eligible: bool = Field(..., examples=[True])
    match_percentage: float = Field(..., examples=[100.0])
    criteria_passed: int = Field(..., examples=[2])
    criteria_total: int = Field(..., examples=[2])
    summary_reason: str = Field(..., examples=["You meet all 2 eligibility criteria for this scheme."])
    passed_criteria: list[CriterionVerdict] = []
    failed_criteria: list[CriterionVerdict] = []
    benefits_summary: list[str] = Field(default=[], examples=[["₹6,000 Annual Direct Cash Benefit"]])
    application_url: str | None = Field(None, examples=["https://pmkisan.gov.in"])


class EligibilityReportResponse(BaseModel):
    total_evaluated: int = Field(..., examples=[12], description="Total active schemes analyzed")
    eligible_count: int = Field(..., examples=[3], description="Schemes with 100% criteria matched")
    nearly_eligible_count: int = Field(..., examples=[2], description="Schemes with 50%-99% criteria matched")
    ineligible_count: int = Field(..., examples=[7], description="Schemes with <50% criteria matched")
    eligible_schemes: list[SchemeExplanation] = Field(default=[], description="List of fully matched schemes")
    nearly_eligible_schemes: list[SchemeExplanation] = Field(default=[], description="List of nearly matched schemes with unmet conditions")
    ineligible_schemes: list[SchemeExplanation] = Field(default=[], description="List of non-matching schemes")

File: scheme-backend-go/internal/features/eligibility/dto/eligibility.go
package dto

type EligibilityCheckRequest struct {
	DateOfBirth        *string `json:"date_of_birth,omitempty"` // YYYY-MM-DD
	Age                *int    `json:"age,omitempty"`
	Gender             *string `json:"gender,omitempty"`
	State              *string `json:"state,omitempty"`
	District           *string `json:"district,omitempty"`
	AnnualIncome       *int32  `json:"annual_income,omitempty"`
	Occupation         *string `json:"occupation,omitempty"`
	CasteCategory      *string `json:"caste_category,omitempty"`
	IsDifferentlyAbled *bool   `json:"is_differently_abled,omitempty"`
	MaritalStatus      *string `json:"marital_status,omitempty"`
	ResidenceArea      *string `json:"residence_area,omitempty"`
	HasLand            *bool   `json:"has_land,omitempty"`
}

type CriterionVerdict struct {
	Field             string `json:"field"`
	CriterionTitle    string `json:"criterion_title"`
	Status            string `json:"status"` // "passed" | "failed" | "missing_info"
	YourValue         string `json:"your_value"`
	RequiredCondition string `json:"required_condition"`
	Reason            string `json:"reason"`
}

type SchemeExplanation struct {
	SchemeID         int32              `json:"scheme_id"`
	SchemeName       string             `json:"scheme_name"`
	SchemeSlug       string             `json:"scheme_slug"`
	State            string             `json:"state"`
	Ministry         string             `json:"ministry"`
	Description      string             `json:"description"`
	Status           string             `json:"status"` // "eligible" | "nearly_eligible" | "ineligible"
	IsEligible       bool               `json:"is_eligible"`
	MatchPercentage  float64            `json:"match_percentage"`
	CriteriaPassed   int                `json:"criteria_passed"`
	CriteriaTotal    int                `json:"criteria_total"`
	SummaryReason    string             `json:"summary_reason"`
	PassedCriteria   []CriterionVerdict `json:"passed_criteria"`
	FailedCriteria   []CriterionVerdict `json:"failed_criteria"`
	BenefitsSummary  []string           `json:"benefits_summary"`
	ApplicationURL   *string            `json:"application_url,omitempty"`
}

type EligibilityReportResponse struct {
	TotalEvaluated        int                 `json:"total_evaluated"`
	EligibleCount         int                 `json:"eligible_count"`
	NearlyEligibleCount   int                 `json:"nearly_eligible_count"`
	IneligibleCount       int                 `json:"ineligible_count"`
	EligibleSchemes       []SchemeExplanation `json:"eligible_schemes"`
	NearlyEligibleSchemes []SchemeExplanation `json:"nearly_eligible_schemes"`
	IneligibleSchemes     []SchemeExplanation `json:"ineligible_schemes"`
}
